### PR TITLE
Set Resources on all Jobs

### PR DESCRIPTION
**Pull request by jmckulk**
This change allows users to define resource requests and limits for all
jobs created by PGO. This includes, manual backup, replica
create backup, restore, and move jobs for pgdata, pgwal and repoHost
volumes.

A new field has been added to the spec to define resource options for
backup jobs. `spec.backups.pgbackrest.jobs.resources` is a Kubernetes
ResourceRequirements field that is passed to all backup jobs.

Resources for move jobs will be set based on the type of volume the job
is working on. pgdata and pgwal volume move jobs will use the resource
field defined for the first instance defined for the cluster (i.e.
`spec.instances[0])`.  repoHost volume move jobs will use resources
defined for the repo host
(`spec.backups.pgbackrest.repoHost.resources`).

Resources for restore jobs are configurable using
`spec.backups.pgbackrest.restore.resources`.

**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [ ] Have you updated or added documentation for the change, as applicable?
 - [x] Have you tested your changes on all related environments with successful results, as applicable?

**Type of Changes:**

 
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**

**What is the new behavior (if this is a feature change)?**

**Other information**:
[sc-12554]

### CODE REVIEW

File: internal/controller/postgrescluster/pgbackrest_test.go [hunk 1]
```diff
@@ -2194,6 +2194,46 @@ func TestReconcilePostgresClusterDataSource(t *testing.T) {
...
+	t.Run("empty", func(t *testing.T) {
+		_, err := generateBackupJobSpecIntent(
+			&v1beta1.PostgresCluster{},
+			"", "", "", "", "",
+			nil, nil,
+		)
+		assert.NilError(t, err)
+	})
+
+	cluster := &v1beta1.PostgresCluster{
```
Comment (jmckulk) [inline internal/controller/postgrescluster/pgbackrest_test.go:2207]: I want to add a test around here to generate the spec when jobs is defined but resources is not. I'll push it up with the docs changes. :recycle:

File: internal/controller/postgrescluster/volumes.go [hunk 2]
```diff
@@ -449,6 +449,7 @@ func (r *Reconciler) reconcileMovePGDataDir(ctx context.Context,
 					Image:        config.PostgresContainerImage(cluster),
 					Name:         naming.ContainerJobMovePGDataDir,
 					VolumeMounts: []corev1.VolumeMount{postgres.DataVolumeMount()},
+					Resources:    cluster.Spec.InstanceSets[0].Resources,
```
Comment (tjmoore4) [inline internal/controller/postgrescluster/volumes.go:452]: :thinking: Is there any chance `InstanceSets` might ever be empty? Perhaps a `len(cluster.Spec.InstanceSets) > 0` check above might be useful?

File: internal/controller/postgrescluster/volumes.go [hunk 3]
```diff
@@ -548,6 +549,7 @@ func (r *Reconciler) reconcileMoveWALDir(ctx context.Context,
 					Image:        config.PostgresContainerImage(cluster),
 					Name:         naming.ContainerJobMovePGWALDir,
 					VolumeMounts: []corev1.VolumeMount{postgres.WALVolumeMount()},
+					Resources:    cluster.Spec.InstanceSets[0].Resources,
```
Comment (tjmoore4) [inline internal/controller/postgrescluster/volumes.go:552]: :thinking: Same as above

File: internal/controller/postgrescluster/volumes.go [hunk 4]
```diff
@@ -652,6 +654,7 @@ func (r *Reconciler) reconcileMoveRepoDir(ctx context.Context,
 					Image:        config.PGBackRestContainerImage(cluster),
 					Name:         naming.ContainerJobMovePGBackRestRepoDir,
 					VolumeMounts: []corev1.VolumeMount{pgbackrest.RepoVolumeMount()},
+					Resources:    cluster.Spec.Backups.PGBackRest.RepoHost.Resources,
```
Comment (tjmoore4) [inline internal/controller/postgrescluster/volumes.go:657]: :wrench:  Similar to the above, but I believe it's likely there are cases where `RepoHost` will be `nil`.

**Review by tjmoore4 — COMMENTED** [review]
Looks good overall. A few notes, and I'll be on the lookout for the mentioned updates.

File: internal/controller/postgrescluster/volumes.go (hunk 2 above)
Comment (jmckulk) [reply internal/controller/postgrescluster/volumes.go:452]: :thinking: I don't think that is possible based on spec validation.. Maybe in a shutdown scenario? Either way would be good to have a check

**Review by tjmoore4 — APPROVED** [review]

File: internal/controller/postgrescluster/pgbackrest_test.go [hunk 5]
```diff
@@ -2194,6 +2194,56 @@ func TestReconcilePostgresClusterDataSource(t *testing.T) {
...
+		job, err := generateBackupJobSpecIntent(
+			cluster,
+			"", "", "", "", "",
+			nil, nil,
+		)
+		assert.NilError(t, err)
+		assert.DeepEqual(t, job.Template.Spec.Containers[0].Resources,
+			corev1.ResourceRequirements{
+				Requests: corev1.ResourceList{
+					corev1.ResourceStorage: resource.MustParse("1Gi"),
```
Comment (cbandy) [inline internal/controller/postgrescluster/pgbackrest_test.go:2241]: ⛏️ A container doesn't normally allow `storage` here.

```
Pod "mycluster-00-xcxl-0" is invalid: [spec.containers[0].resources.requests[storage]: Invalid value: "storage": must be a standard resource for containers
```

**Review by cbandy — APPROVED** [review]

**Pull request merged**